src/data_type.hpp

```diff
@@ -452,18 +435,37 @@ class UserType : public DataType {
   CaseInsensitiveHashTable<Field> fields_;
 };
 
-class NativeDataTypes {
+class ValueTypes {
 public:
-  void init_class_names();
-  const DataType::ConstPtr& by_class_name(const std::string& name) const;
+  ValueTypes();
+
+  static CassValueType by_class(StringRef name);
+  static CassValueType by_cql(StringRef name);
+
+private:
+  typedef SmallDenseHashMap<StringRef, CassValueType,
+                            CASS_VALUE_TYPE_LAST_ENTRY, // Max size
+                            StringRefIHash,
+                            StringRefIEquals> HashMap;
+
+  static HashMap value_types_by_class_;
+  static HashMap value_types_by_cql_;
+};
+
+class SimpleDataTypeCache {
+public:
+  const DataType::ConstPtr& by_class(StringRef name) {
```

Comment:
The problem is the external API and other places in the code treat more complex and unique data types (collections, UDTs, tuples, etc.) as generic `DataType`.